File: src/main/scala/akka/stream/contrib/FileTailSource.scala
/*
 * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
 */
package akka.stream.contrib

import java.nio.ByteBuffer
import java.nio.channels.{AsynchronousFileChannel, CompletionHandler}
import java.nio.file.{Files, Path, StandardOpenOption}

import akka.NotUsed
import akka.stream.stage._
import akka.stream.{Attributes, Outlet, SourceShape}
import akka.util.ByteString

import scala.concurrent.duration.FiniteDuration
import scala.util.{Failure, Success, Try}

/**
 * Read the entire contents of a file, and then when the end is reached, keep reading
 * newly appended data. Like the unix command `tail -f`.
 *
 * Aborting the stage can be done by combining with a [[akka.stream.KillSwitch]]
 *
 * @param path a file path to tail
 * @param maxChunkSize The max emitted size of the `ByteString`s
 * @param startingPosition Offset into the file to start reading
 * @param pollingInterval When the end has been reached, look for new content with this interval
 */
final class FileTailSource(path: Path, maxChunkSize: Int, startingPosition: Long, pollingInterval: FiniteDuration) extends GraphStage[SourceShape[ByteString]] {
  val out = Outlet[ByteString]("FileTailSource.out")
  override val shape: SourceShape[ByteString] = SourceShape(out)

  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
    new TimerGraphStageLogic(shape) with OutHandler {
      assert(Files.isReadable(path), s"path $path is not readable")
      implicit def ec = materializer.executionContext

      val buffer = ByteBuffer.allocate(maxChunkSize)
      val channel = AsynchronousFileChannel.open(path, StandardOpenOption.READ)
      var position = startingPosition
      var chunkHandler: CompletionHandler[Integer, NotUsed] = _

      override def preStart(): Unit = {
        val chunkCallback = getAsyncCallback[Try[Integer]] {
          case Success(readBytes) =>
            if (readBytes > 0) {
              buffer.flip()
              push(out, ByteString.fromByteBuffer(buffer))
              position += readBytes
              buffer.clear()
            } else {
              // hit end, try again in a while
              scheduleOnce(NotUsed, pollingInterval)
            }

          case Failure(ex) =>
            failStage(ex)
        }

        chunkHandler = new CompletionHandler[Integer, NotUsed] {
          override def completed(result: Integer, attachment: NotUsed): Unit = {
            chunkCallback.invoke(Success(result))
          }
          override def failed(ex: Throwable, attachment: NotUsed): Unit = {
            chunkCallback.invoke(Failure(ex))
          }
        }
      }

      override protected def onTimer(timerKey: Any): Unit = {
        onPull()
      }

      override def onPull(): Unit = {
        channel.read(buffer, position, NotUsed, chunkHandler)
      }

      override def postStop(): Unit = {
        if (channel.isOpen) channel.close()
      }

      setHandler(out, this)
    }
}

File: contrib/src/main/java/akka/stream/contrib/FileTailSource.java
/*
 * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
 */
package akka.stream.contrib;

import akka.NotUsed;
import akka.stream.Attributes;
import akka.stream.Outlet;
import akka.stream.SourceShape;
import akka.stream.javadsl.Source;
import akka.stream.stage.*;
import akka.util.ByteString;
import scala.concurrent.duration.FiniteDuration;
import scala.util.Failure;
import scala.util.Success;
import scala.util.Try;

import java.nio.ByteBuffer;
import java.nio.channels.AsynchronousFileChannel;
import java.nio.channels.CompletionHandler;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;

/**
 * Read the entire contents of a file, and then when the end is reached, keep reading
 * newly appended data. Like the unix command `tail -f`.
 *
 * Aborting the stage can be done by combining with a [[akka.stream.KillSwitch]]
 */
public final class FileTailSource extends GraphStage<SourceShape<ByteString>> {

  private final Path path;
  private final int maxChunkSize;
  private final long startingPosition;
  private final FiniteDuration pollingInterval;
  private final Outlet<ByteString> out = Outlet.create("FileTailSource.out");
  private final SourceShape<ByteString> shape = SourceShape.of(out);

  // this is stateless, so can be shared among instances
  private static final CompletionHandler<Integer, AsyncCallback<Try<Integer>>> completionHandler = new CompletionHandler<Integer, AsyncCallback<Try<Integer>>>() {
    @Override
    public void completed(Integer result, AsyncCallback<Try<Integer>> attachment) {
      attachment.invoke(new Success<>(result));
    }

    @Override
    public void failed(Throwable exc, AsyncCallback<Try<Integer>> attachment) {
      attachment.invoke(new Failure<>(exc));
    }
  };

  public FileTailSource(Path path, int maxChunkSize, long startingPosition, FiniteDuration pollingInterval) {
    this.path = path;
    this.maxChunkSize = maxChunkSize;
    this.startingPosition = startingPosition;
    this.pollingInterval = pollingInterval;
  }

  @Override
  public SourceShape<ByteString> shape() {
    return shape;
  }

  @Override
  public GraphStageLogic createLogic(Attributes inheritedAttributes) {
    try {
      if (!Files.exists(path)) throw new IllegalArgumentException("Path '" + path + "' does not exist");
      if (Files.isDirectory(path)) throw new IllegalArgumentException("Path '" + path + "' cannot be tailed, it is a directory");
      if (!Files.isReadable(path)) throw new IllegalArgumentException("No read permission for '" + path + "'");

      return new TimerGraphStageLogic(shape) {
        private final ByteBuffer buffer = ByteBuffer.allocate(maxChunkSize);
        private final AsynchronousFileChannel channel = AsynchronousFileChannel.open(path, StandardOpenOption.READ);

        private long position = startingPosition;
        private AsyncCallback<Try<Integer>> chunkCallback;

        {
          setHandler(out, new AbstractOutHandler() {
            @Override
            public void onPull() throws Exception {
              doPull();
            }
          });
        }

        @Override
        public void preStart() {
          chunkCallback = createAsyncCallback((tryInteger) -> {
            if (tryInteger.isSuccess()) {
              int readBytes = tryInteger.get();
              if (readBytes > 0) {
                buffer.flip();
                push(out, ByteString.fromByteBuffer(buffer));
                position += readBytes;
                buffer.clear();
              } else {
                // hit end, try again in a while
                scheduleOnce("poll", pollingInterval);
              }

            } else {
              failStage(tryInteger.failed().get());
            }

          });
        }

        @Override
        public void onTimer(Object timerKey) {
          doPull();
        }


        private void doPull() {
          channel.read(buffer, position, chunkCallback, completionHandler);
        }

        @Override
        public void postStop() {
          try {
            if (channel.isOpen()) channel.close();
          } catch(Exception ex) {
            // Remove when #21168 is fixed
            throw new RuntimeException(ex);
          }
        }
      };

    } catch (Exception ex) {
      // remove when #21168 is fixed
      throw new RuntimeException(ex);
    }
  }


  // factory methods

  /**
   * Java API:
   *
   * Read the entire contents of a file, and then when the end is reached, keep reading
   * newly appended data. Like the unix command `tail -f`.
   *
   * Aborting the stage can be done by combining with a [[akka.stream.KillSwitch]]
   *
   * @param path a file path to tail
   * @param maxChunkSize The max emitted size of the `ByteString`s
   * @param startingPosition Offset into the file to start reading
   * @param pollingInterval When the end has been reached, look for new content with this interval
   */
  public static Source<ByteString, NotUsed> create(Path path, int maxChunkSize, long startingPosition, FiniteDuration pollingInterval) {
    return Source.fromGraph(new FileTailSource(path, maxChunkSize, startingPosition, pollingInterval));
  }

  /**
   * Scala API:
   *
   * Read the entire contents of a file, and then when the end is reached, keep reading
   * newly appended data. Like the unix command `tail -f`.
   *
   * Aborting the stage can be done by combining with a [[akka.stream.KillSwitch]]
   *
   * @param path a file path to tail
   * @param maxChunkSize The max emitted size of the `ByteString`s
   * @param startingPosition Offset into the file to start reading
   * @param pollingInterval When the end has been reached, look for new content with this interval
   */
  public static akka.stream.scaladsl.Source<ByteString, NotUsed> apply(Path path, int maxChunkSize, long startingPosition, FiniteDuration pollingInterval) {
    return create(path, maxChunkSize, startingPosition, pollingInterval).asScala();
  }

}
